### ISSUE TITLE
catkin_make error 'undefined reference to...'

### ISSUE THREAD

**Issue by MorganLeee** [atenpas/gpd#25]
Following the instruction and catkin_make error occurs:
`morgan@morgan-Z170X:~/cute_ws$ catkin_make
Base path: /home/morgan/cute_ws
Source space: /home/morgan/cute_ws/src
Build space: /home/morgan/cute_ws/build
Devel space: /home/morgan/cute_ws/devel
Install space: /home/morgan/cute_ws/install
####
####
####
####
...
[ 94%] Linking CXX executable /home/morgan/cute_ws/devel/lib/gpd/test_occlusion
[ 95%] Built target gpd_sequential_importance_sampling
[ 96%] Linking CXX executable /home/morgan/cute_ws/devel/lib/gpd/create_training_data
[ 96%] Linking CXX executable /home/morgan/cute_ws/devel/lib/gpd/detect_grasps
[ 96%] Linking CXX executable /home/morgan/cute_ws/devel/lib/gpd/classify_candidates
/home/morgan/cute_ws/devel/lib/libgpd_data_generator.so: undefined reference to `caffe::db::GetDB(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/home/morgan/cute_ws/devel/lib/libgpd_data_generator.so: undefined reference to `caffe::Datum::Datum()'
/home/morgan/cute_ws/devel/lib/libgpd_data_generator.so: undefined reference to `google::LogMessage::stream()'
/home/morgan/cute_ws/devel/lib/libgpd_data_generator.so: undefined reference to `google::protobuf::MessageLite::SerializeToString(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*) const'
/home/morgan/cute_ws/devel/lib/libgpd_data_generator.so: undefined reference to `google::LogMessage::LogMessage(char const*, int)'
/home/morgan/cute_ws/devel/lib/libgpd_data_generator.so: undefined reference to `caffe::CVMatToDatum(cv::Mat const&, caffe::Datum*)'
/home/morgan/cute_ws/devel/lib/libgpd_data_generator.so: undefined reference to `caffe::Datum::~Datum()'
/home/morgan/cute_ws/devel/lib/libgpd_data_generator.so: undefined reference to `google::LogMessageFatal::~LogMessageFatal()'
/home/morgan/cute_ws/devel/lib/libgpd_data_generator.so: undefined reference to `google::LogMessage::~LogMessage()'
/home/morgan/cute_ws/devel/lib/libgpd_data_generator.so: undefined reference to `google::LogMessageFatal::LogMessageFatal(char const*, int)'
collect2: error: ld returned 1 exit status
gpd/CMakeFiles/gpd_create_training_data.dir/build.make:394: recipe for target '/home/morgan/cute_ws/devel/lib/gpd/create_training_data' failed
make[2]: *** [/home/morgan/cute_ws/devel/lib/gpd/create_training_data] Error 1
CMakeFiles/Makefile2:2568: recipe for target 'gpd/CMakeFiles/gpd_create_training_data.dir/all' failed
make[1]: *** [gpd/CMakeFiles/gpd_create_training_data.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
/tmp/ccQCJ29Z.ltrans0.ltrans.o: In function `main':
<artificial>:(.text.startup+0x6f6): undefined reference to `caffe::Caffe::Get()'
<artificial>:(.text.startup+0x7ad): undefined reference to `caffe::Net<float>::Net(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, caffe::Phase, int, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const*)'
<artificial>:(.text.startup+0x7e7): undefined reference to `caffe::Net<float>::CopyTrainedLayersFrom(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'
<artificial>:(.text.startup+0x8a2): undefined reference to `caffe::Net<float>::layer_by_name(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
<artificial>:(.text.startup+0xccc): undefined reference to `caffe::Net<float>::Forward(float*)'
<artificial>:(.text.startup+0xdae): undefined reference to `caffe::Net<float>::blob_by_name(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
<artificial>:(.text.startup+0xe76): undefined reference to `caffe::Blob<float>::cpu_data() const'
<artificial>:(.text.startup+0x1312): undefined reference to `google::LogMessageFatal::LogMessageFatal(char const*, int, google::CheckOpString const&)'
<artificial>:(.text.startup+0x131a): undefined reference to `google::LogMessage::stream()'
<artificial>:(.text.startup+0x1336): undefined reference to `google::LogMessageFatal::~LogMessageFatal()'
<artificial>:(.text.startup+0x163b): undefined reference to `google::LogMessageFatal::LogMessageFatal(char const*, int, google::CheckOpString const&)'
<artificial>:(.text.startup+0x1643): undefined reference to `google::LogMessage::stream()'
<artificial>:(.text.startup+0x165f): undefined reference to `google::LogMessageFatal::~LogMessageFatal()'
<artificial>:(.text.startup+0x18ce): undefined reference to `google::LogMessageFatal::~LogMessageFatal()'
<artificial>:(.text.startup+0x1a51): undefined reference to `google::LogMessageFatal::~LogMessageFatal()'
/tmp/ccQCJ29Z.ltrans5.ltrans.o: In function `caffe::Blob<float>::LegacyShape(int) const [clone .part.233] [clone .constprop.14]':
<artificial>:(.text+0xa4a): undefined reference to `google::LogMessageFatal::LogMessageFatal(char const*, int, google::CheckOpString const&)'
<artificial>:(.text+0xa52): undefined reference to `google::LogMessage::stream()'
<artificial>:(.text+0xac8): undefined reference to `google::LogMessageFatal::~LogMessageFatal()'
<artificial>:(.text+0xc7b): undefined reference to `google::LogMessageFatal::LogMessageFatal(char const*, int, google::CheckOpString const&)'
<artificial>:(.text+0xc83): undefined reference to `google::LogMessage::stream()'
<artificial>:(.text+0xcf9): undefined reference to `google::LogMessageFatal::~LogMessageFatal()'
<artificial>:(.text+0xf98): undefined reference to `google::LogMessageFatal::~LogMessageFatal()'
<artificial>:(.text+0xfa0): undefined reference to `google::LogMessageFatal::~LogMessageFatal()'
/tmp/ccQCJ29Z.ltrans8.ltrans.o: In function `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >* google::MakeCheckOpString<int, int>(int const&, int const&, char const*)':
<artificial>:(.text+0x878): undefined reference to `google::base::CheckOpMessageBuilder::CheckOpMessageBuilder(char const*)'
<artificial>:(.text+0x88d): undefined reference to `google::base::CheckOpMessageBuilder::ForVar2()'
<artificial>:(.text+0x8a0): undefined reference to `google::base::CheckOpMessageBuilder::NewString[abi:cxx11]()'
<artificial>:(.text+0x8ab): undefined reference to `google::base::CheckOpMessageBuilder::~CheckOpMessageBuilder()'
<artificial>:(.text+0x8c2): undefined reference to `google::base::CheckOpMessageBuilder::~CheckOpMessageBuilder()'
collect2: error: ld returned 1 exit status
gpd/CMakeFiles/gpd_test_occlusion.dir/build.make:453: recipe for target '/home/morgan/cute_ws/devel/lib/gpd/test_occlusion' failed
make[2]: *** [/home/morgan/cute_ws/devel/lib/gpd/test_occlusion] Error 1
CMakeFiles/Makefile2:3265: recipe for target 'gpd/CMakeFiles/gpd_test_occlusion.dir/all' failed
make[1]: *** [gpd/CMakeFiles/gpd_test_occlusion.dir/all] Error 2
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::base::CheckOpMessageBuilder::CheckOpMessageBuilder(char const*)'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::base::CheckOpMessageBuilder::~CheckOpMessageBuilder()'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `caffe::Net<float>::Net(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, caffe::Phase, int, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const*)'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::base::CheckOpMessageBuilder::ForVar2()'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `caffe::Blob<float>::cpu_data() const'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::LogMessage::stream()'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `caffe::Net<float>::CopyTrainedLayersFrom(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::LogMessageFatal::LogMessageFatal(char const*, int, google::CheckOpString const&)'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `caffe::Caffe::Get()'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `caffe::Net<float>::layer_by_name(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::LogMessageFatal::~LogMessageFatal()'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `caffe::Net<float>::Forward(float*)'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::base::CheckOpMessageBuilder::NewString[abi:cxx11]()'
collect2: error: ld returned 1 exit status
gpd/CMakeFiles/gpd_detect_grasps.dir/build.make:440: recipe for target '/home/morgan/cute_ws/devel/lib/gpd/detect_grasps' failed
make[2]: *** [/home/morgan/cute_ws/devel/lib/gpd/detect_grasps] Error 1
CMakeFiles/Makefile2:2251: recipe for target 'gpd/CMakeFiles/gpd_detect_grasps.dir/all' failed
make[1]: *** [gpd/CMakeFiles/gpd_detect_grasps.dir/all] Error 2
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::base::CheckOpMessageBuilder::CheckOpMessageBuilder(char const*)'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::base::CheckOpMessageBuilder::~CheckOpMessageBuilder()'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `caffe::Net<float>::Net(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, caffe::Phase, int, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const*)'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::base::CheckOpMessageBuilder::ForVar2()'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `caffe::Blob<float>::cpu_data() const'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::LogMessage::stream()'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `caffe::Net<float>::CopyTrainedLayersFrom(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::LogMessageFatal::LogMessageFatal(char const*, int, google::CheckOpString const&)'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `caffe::Caffe::Get()'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `caffe::Net<float>::layer_by_name(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::LogMessageFatal::~LogMessageFatal()'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `caffe::Net<float>::Forward(float*)'
/home/morgan/cute_ws/devel/lib/libgpd_caffe_classifier.so: undefined reference to `google::base::CheckOpMessageBuilder::NewString[abi:cxx11]()'
collect2: error: ld returned 1 exit status
gpd/CMakeFiles/gpd_classify_candidates.dir/build.make:461: recipe for target '/home/morgan/cute_ws/devel/lib/gpd/classify_candidates' failed
make[2]: *** [/home/morgan/cute_ws/devel/lib/gpd/classify_candidates] Error 1
CMakeFiles/Makefile2:3416: recipe for target 'gpd/CMakeFiles/gpd_classify_candidates.dir/all' failed
make[1]: *** [gpd/CMakeFiles/gpd_classify_candidates.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
Invoking "make -j8 -l8" failed
`
Thanks for attention!

**Comment by MorganLeee** [comment]
sloved

**Comment by wkoa** [comment]
How had you solved this?? can you tell me?  i  meet the same problem now

**Comment by mfinean** [comment]
What did you change when you reinstalled? I have the same issue

**Comment by mfinean** [comment]
In case anyone has this error - it's basically because a few executables aren't linking correctly to the C++ definitions. The root of my problem was that I built Caffe from source and as such needed to have a FindCaffe.cmake for finding Caffe. In doing this I set Caffe_LIBS to be the libraries to add in the CMakeLists.txt file. But later on classifier_dep is set using Caffe_LIBRARIES, which was empty in my case. Make sure that classifier_dep is set correctly! All the best
